auth/src/metadata.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use http::HeaderMap;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use std::env;
+use tokio::time::{self, Duration};
+
+use super::{Error, Result};
+
+// TODO(codyoss): cache a client
+// TODO(codyoss): move client impl elsewhere
+// TODO(codyoss): Instead of vecs things could be slices... don't need ownership
+// TODO(codyoss): funcs could take &str or impl Into<String>?
+
+/// A wrapper around a HTTP client used to talk to a Google Cloud metadata service.
+struct MetadataClient {
+    c: Client,
+    // TODO(codyoss): could include a cache here
+}
+
+impl MetadataClient {
+    /// Create a new Metadata client with the proper headers required to talk to
+    /// the Google Cloud metadata service.
+    fn new() -> Self {
+        let mut headers = HeaderMap::with_capacity(2);
+        headers.insert("Metadata-Flavor", "Google".parse().unwrap());
+        headers.insert("User-Agent", "gcloud-rust/0.1".parse().unwrap());
+        let client = Client::builder().default_headers(headers).build().unwrap();
+        Self { c: client }
+    }
+}
+
+/// Makes a request to the supplied metadata endpoint.
+///
+/// ```rust
+/// let project_id = get("project/project-id")?;
+/// println!("{}", project_id);
+/// ```
+// TODO(codyoss): I wonder if this should be a pathbuf
+#[allow(dead_code)]
+pub async fn get(suffix: String) -> Result<String> {
+    get_with_query::<()>(suffix, None).await
+}
+
+async fn get_with_query<T: Serialize + ?Sized>(
+    suffix: String,
+    query: Option<&T>,
+) -> Result<String> {
+    let host = env::var("GCE_METADATA_HOST")
```

Comment:
Please insert a comment with a link to where this environment variable and IP address are documented.